### PR TITLE
Introduce pipeline parallel degree config

### PR DESCRIPTION
**Pull request by nskool**
## Description ##

Brief description of what this PR is about

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

File: engines/python/setup/djl_python/rolling_batch/lmi_dist_rolling_batch.py [hunk 1]
```diff
@@ -81,6 +82,8 @@ def __init__(self, model_id_or_path: str, properties: dict, **kwargs):
             **engine_kwargs)
 
         kwargs = {}
+        print(f"engine_args: {engine_args}, kwargs={kwargs}")
```
Comment (frankfliu) [inline engines/python/setup/djl_python/rolling_batch/lmi_dist_rolling_batch.py:85]: use logging

File: engines/python/src/main/java/ai/djl/python/engine/Connection.java [hunk 2]
```diff
@@ -126,12 +126,23 @@ static String[] getPythonStartCmd(
...
         int clusterSize = PyEnv.getClusterSize();
         int tensorParallelDegree = pyEnv.getTensorParallelDegree();
+        int pipelineParallelDegree = pyEnv.getPipelineParallelDegree();
         String entryPoint = pyEnv.getEntryPoint();
         String recommendedEntryPoint = pyEnv.getRecommendedEntryPoint();
 
         if (PyEnv.isMultiNode()) {
-            String cudaDevices = getVisibleDevices(workerId, tensorParallelDegree / clusterSize);
+
+            int mpiSlots = (tensorParallelDegree*pipelineParallelDegree) / 2;
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/Connection.java:135]: Why divide by 2?

`tensorParallelDegree*pipelineParallelDegree / clusterSize`?

File: engines/python/src/main/java/ai/djl/python/engine/Connection.java [hunk 3]
```diff
@@ -126,12 +126,23 @@ static String[] getPythonStartCmd(
...
+            int mpiProcesses = tensorParallelDegree*pipelineParallelDegree;
+
+            // if (pipelineParallelDegree != clusterSize) {
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/Connection.java:138]: Remove commented out code.
We should check if total slots is divisible by clusterSize

File: engines/python/src/main/java/ai/djl/python/engine/Connection.java [hunk 4]
```diff
@@ -140,12 +151,12 @@ static String[] getPythonStartCmd(
...
                 }
-                sb.append(host).append(':').append(tensorParallelDegree / clusterSize);
+                sb.append(host).append(':').append(mpiSlots);
             }
-            String[] args = new String[46];
+            String[] args = new String[48];
             args[0] = "mpirun";
             args[1] = "-np";
-            args[2] = String.valueOf(tensorParallelDegree);
+            args[2] = String.valueOf(mpiProcesses);
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/Connection.java:159]: Better use `worldSize`

File: engines/python/src/main/java/ai/djl/python/engine/Connection.java [hunk 5]
```diff
@@ -126,12 +126,23 @@ static String[] getPythonStartCmd(
...
+
+            int mpiSlots = (tensorParallelDegree*pipelineParallelDegree) / 2;
+            int mpiProcesses = tensorParallelDegree*pipelineParallelDegree;
+
+            // if (pipelineParallelDegree != clusterSize) {
+            //     logger.warn("In multi-node setting, pipeline parallel degree must equal the cluster size. Setting pp degree = cluster size = {}.", clusterSize);
+            //     pipelineParallelDegree = clusterSize;
+            // }
+
+            String cudaDevices = getVisibleDevices(workerId, mpiSlots);
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/Connection.java:143]: rename `mpiSlots` to `localSize`

File: engines/python/setup/djl_python/properties_manager/hf_properties.py [hunk 6]
```diff
@@ -120,14 +121,15 @@ def construct_kwargs_device_map(self):
...
+                and self.pipeline_parallel_degree > 0 \
                 and self.cluster_size > 0 \
                 and torch.cuda.device_count() > 0:
             self.kwargs["device_map"] = "auto"
             self.device = None
             world_size = torch.cuda.device_count() * self.cluster_size
-            assert world_size == self.tensor_parallel_degree, \
-                f"TP degree ({self.tensor_parallel_degree}) doesn't match available GPUs ({world_size})"
-            logging.info(f"Using {world_size} gpus")
+            assert world_size == self.tensor_parallel_degree*self.pipeline_parallel_degree, \
```
Comment (frankfliu) [inline engines/python/setup/djl_python/properties_manager/hf_properties.py:130]: avoid use `assert` in production code:

```
            if world_size != self.tensor_parallel_degree * self.pipeline_parallel_degree:
                raise ValueError(
                    f"TP*PP degree ({self.tensor_parallel_degree*self.pipeline_parallel_degree}) doesn't match available GPUs ({world_size})"
                )
```

Comment (nskool) [reply engines/python/setup/djl_python/properties_manager/hf_properties.py]: I did not add this assert previously, just modified it. I have now removed it to use ValueError

**Review by frankfliu — APPROVED** [review]

**Pull request merged**